src/librustdoc/clean/types.rs

```diff
@@ -641,6 +643,55 @@ impl Item {
             _ => false,
         }
     }
+
+    /// Returns the `HirId` of `self`. It'll panic if `self` doesn't have a local `DefId`!
+    crate fn hir_id(&self) -> hir::HirId {
+        hir::HirId::make_owner(
+            self.def_id.as_def_id().unwrap().as_local().expect("not local DefId"),
+        )
+    }
+
+    /// Returns a `FnHeader` if `self` is a function item, otherwise returns `None`.
+    crate fn fn_header(&self, tcx: TyCtxt<'_>) -> Option<hir::FnHeader> {
+        let header = match *self.kind {
+            ItemKind::ForeignFunctionItem(_) => {
+                let abi = tcx.hir().get_foreign_abi(self.hir_id());
+                hir::FnHeader {
+                    unsafety: if abi == Abi::RustIntrinsic {
+                        intrinsic_operation_unsafety(self.name.unwrap())
+                    } else {
+                        hir::Unsafety::Unsafe
+                    },
+                    abi,
+                    constness: hir::Constness::NotConst,
+                    asyncness: hir::IsAsync::NotAsync,
+                }
+            }
+            ItemKind::FunctionItem(_) | ItemKind::MethodItem(_, _) => {
+                let def_id = self.def_id.as_def_id().unwrap();
+                let sig = tcx.fn_sig(def_id);
+                let constness =
+                    if tcx.is_const_fn(def_id) && is_unstable_const_fn(tcx, def_id).is_none() {
+                        hir::Constness::Const
+                    } else {
+                        hir::Constness::NotConst
+                    };
+                let asyncness = tcx.asyncness(def_id);
+                hir::FnHeader { unsafety: sig.unsafety(), abi: sig.abi(), constness, asyncness }
+            }
+            ItemKind::TyMethodItem(_) => {
+                let sig = tcx.fn_sig(self.def_id.as_def_id().unwrap());
+                hir::FnHeader {
+                    unsafety: sig.unsafety(),
+                    abi: sig.abi(),
+                    constness: hir::Constness::NotConst,
+                    asyncness: hir::IsAsync::NotAsync,
+                }
+            }
+            _ => return None,
+        };
+        Some(header)
+    }
```

Comment:
Can you use `tcx.fn_sig()` everywhere here and get rid of all the HIR stuff (including the `hir_id()` method)? I don't think we need to rely on HIR to get ABI info.

Reply:
You were perfectly right! Thanks for the suggestion. The code is now smaller. :)